### DISCUSSION TITLE
[Feature Request] Multi-Phrase Filtering with Alternate Definitions

### DISCUSSION THREAD

**Discussion by Psykatte** [mastodon/mastodon#20282] [General]
The ability to add multiple phrases with multiple definitions to a filter rule would allow improved context-sensitive filtering.

In example:
`(cats, dogs)` and `(fish, snails)`

Expands to:
- `cats` and `fish`
- `cats` and `snails`
- `dogs` and `fish`
- `dogs` and `snails`

Effectively filtering all posts that contain the first and last phrase from within one rule.

Here's an example of how this would look:
![filter_example](https://user-images.githubusercontent.com/82907907/200963301-9f33fe3d-ab45-4fcb-b485-6667d42f42a4.png)

It could be deserialized as:
```json
{
  "id": "8449",
  "phrase": [
    [
      "cats",
      "dogs"
    ],
    [
      "fish",
      "snails"
    ]
  ],
  "context": [
    "home",
    "public",
  ],
  "whole_word": false,
  "expires_at": null,
  "irreversible": true
}
```

**Comment by evenprimes** [comment]
This is kind of a combination/generalization of the following:

* [Allow filters to target specific users. #22528](https://github.com/mastodon/mastodon/discussions/22528)
* [Separated filters for lists and and timeline #22073](https://github.com/mastodon/mastodon/discussions/22073)
* [Filtering Variables #19956](https://github.com/mastodon/mastodon/discussions/19956)

I'd like to have this kind of complex filter so that I could do things like:
"Filter applies to the #mastoart hastag, but filter out posts with no image or that include the #trauma tag"
